zerver/lib/bot_lib.py

```diff
@@ -69,6 +71,10 @@ class EmbeddedBotEmptyRecipientsList(Exception):
     pass
 
 
+class EmbeddedBotKeyNotFound(Exception):
+    pass
```

Comment:
This can be a followup PR, but I think structurally we probably want these to be all subclasses of some exception with common exception handling logic that does something reasonable.

(I believe all of these would represent bugs in the embedded bot, so should log an error for server administrators to investigate)

Reply:
Yes. We can do `EmbeddedBotError` or something here. I think we should raise exceptions in a similar manner for external bots as well.

Reply:
I think we can create a protocol for any bot errors that can be applied to both embedded bots and external bots. For example, when the user is writing a bot, they will expect `update_message` to throw an error instead of returning a dictionary that looks like:
```
{
    "code": "BAD_REQUEST",
    "msg": "You don't have permission to edit this message",
    "result": "error"
}
```
Instead, we may have something that `BotError` that applies to both external bots and embedded bots. This will allow better error handling practice.

Reply:
Since there is an expectation that a dict should be returned here, should we return a dict in the cases where an exception is raised at all?

Reply:
That's quite a good question, but we can't do both of them at the same time. I think the ideal case for us might be migrating to use these exceptions before making the error handling logic final. As after all, we would like to replace the dictionaries by exceptions (this part of the work might need to be completed on the python-zulip-api part.

Reply:
But speaking of compatibility, perhaps we can postpone adding the exceptions for now?

Reply:
Hmm, maybe, I am honestly not sure! My initial sense is that just raising an exception without letting the user know anything kinda breaks expectations a little. But yes, we can probably postpone the work on exceptions and do it in a follow-up PR.